pkg/provider/infisical/provider.go

```diff
@@ -0,0 +1,144 @@
+	EnvironmentSlug string
+}
+
+// https://github.com/external-secrets/external-secrets/issues/644
+var _ esv1beta1.SecretsClient = &Provider{}
+var _ esv1beta1.Provider = &Provider{}
+
+func init() {
+	esv1beta1.Register(&Provider{}, &esv1beta1.SecretStoreProvider{
+		Infisical: &esv1beta1.InfisicalProvider{},
+	})
+}
+
+func (p *Provider) Capabilities() esv1beta1.SecretStoreCapabilities {
+	return esv1beta1.SecretStoreReadOnly
+}
+
+func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore, kube kclient.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	storeSpec := store.GetSpec()
+
+	if storeSpec == nil || storeSpec.Provider == nil || storeSpec.Provider.Infisical == nil {
+		return nil, errors.New("invalid infisical store")
+	}
+
+	infisicalSpec := storeSpec.Provider.Infisical
+
+	apiClient, err := api.NewAPIClient(infisicalSpec.HostAPI)
+	if err != nil {
+		return nil, err
+	}
+
+	if infisicalSpec.Auth.Type == constants.UniversalAuth {
+		universalAuthCredentials := infisicalSpec.Auth.UniversalAuthCredentials
+		clientID, err := GetSecretData(ctx, store, kube, namespace, universalAuthCredentials.ClientID)
+		if err != nil {
+			return nil, err
+		}
+
+		clientSecret, err := GetSecretData(ctx, store, kube, namespace, universalAuthCredentials.ClientSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		tokenManager := NewMachineTokenManger(apiClient, clientID, clientSecret)
+		apiClient.SetTokenManager(tokenManager)
+
+		if err := tokenManager.StartLifeCycle(); err != nil {
+			return nil, err
+		}
+
+		return &Provider{
+			apiClient: apiClient,
+			apiScope: &InfisicalClientScope{
+				SecretPath:      infisicalSpec.SecretsScope.SecretsPath,
+				ProjectSlug:     infisicalSpec.SecretsScope.ProjectSlug,
+				EnvironmentSlug: infisicalSpec.SecretsScope.EnvironmentSlug,
+			},
+		}, nil
+	}
+
+	return &Provider{}, errors.New("authentication method not found")
+}
+
+func GetSecretData(ctx context.Context, store esv1beta1.GenericStore, kube kclient.Client, namespace string, secret esmeta.SecretKeySelector) (string, error) {
+	secretRef := esmeta.SecretKeySelector{
+		Name: secret.Name,
+		Key:  secret.Key,
+	}
+	if secret.Namespace != nil {
+		secretRef.Namespace = secret.Namespace
+	}
+
+	secretData, err := resolvers.SecretKeyRef(ctx, kube, store.GetObjectKind().GroupVersionKind().Kind, namespace, &secretRef)
+	if err != nil {
+		return "", err
+	}
+	return secretData, nil
+}
+
+func (p *Provider) ValidateStore(store esv1beta1.GenericStore) (admission.Warnings, error) {
+	storeSpec := store.GetSpec()
+	infisicalStoreSpec := storeSpec.Provider.Infisical
+
+	if infisicalStoreSpec.SecretsScope.EnvironmentSlug == "" || infisicalStoreSpec.SecretsScope.ProjectSlug == "" {
+		return nil, errors.New("secretsScope.projectSlug and secretsScope.environmentSlug cannot be empty")
+	}
+
+	if infisicalStoreSpec.Auth.Type == constants.UniversalAuth {
+		uaCredential := infisicalStoreSpec.Auth.UniversalAuthCredentials
```

Comment:
Likewise - we need to check if the pointer is `nil` or not - otherwise panic when trying to fetch objects from it.